bundles/org.openhab.binding.upb/src/main/java/org/openhab/binding/upb/handler/SerialIoThread.java

```diff
@@ -0,0 +1,297 @@
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.upb.handler;
+
+import static java.nio.charset.StandardCharsets.US_ASCII;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Arrays;
+import java.util.TooManyListenersException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.common.NamedThreadFactory;
+import org.eclipse.smarthome.io.transport.serial.SerialPort;
+import org.eclipse.smarthome.io.transport.serial.SerialPortEvent;
+import org.eclipse.smarthome.io.transport.serial.SerialPortEventListener;
+import org.openhab.binding.upb.handler.UPBIoHandler.CmdStatus;
+import org.openhab.binding.upb.internal.message.MessageBuilder;
+import org.openhab.binding.upb.internal.message.UPBMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Event loop for serial communications. Handles sending and receiving UPB messages.
+ *
+ * @author Marcus Better - Initial contribution
+ *
+ */
+@NonNullByDefault
+public class SerialIoThread extends Thread implements SerialPortEventListener {
+    private static final int WRITE_QUEUE_LENGTH = 128;
+    private static final int ACK_TIMEOUT_MS = 500;
+    private static final byte[] ENABLE_MESSAGE_MODE_CMD = { 0x17, 0x70, 0x02, (byte) 0x8e, 0x0d };
+
+    private final Logger logger = LoggerFactory.getLogger(SerialIoThread.class);
+    private final byte[] buffer = new byte[512];
+    private final MessageListener listener;
+    // Single-threaded executor for writes that serves to serialize writes.
+    private final ExecutorService writeExecutor = new ThreadPoolExecutor(1, 1, 30, TimeUnit.SECONDS,
+            new LinkedBlockingQueue<>(WRITE_QUEUE_LENGTH), new NamedThreadFactory("upb-serial-writer", true));
+    private final SerialPort serialPort;
+
+    private int bufferLength = 0;
+    private volatile @Nullable WriteRunnable currentWrite;
+    private volatile boolean done;
+
+    public SerialIoThread(final SerialPort serialPort, final MessageListener listener) {
+        this.serialPort = serialPort;
+        this.listener = listener;
+        setName("upb-serial-reader");
```

Comment:
Could you use the following naming convention for the thread:
```suggestion
        setName("OH-binding-" + thingUID + "serial-reader");
```
The postfix `serial-reader` is maybe not needed it this is the only thread started by this binding. The convention is to easily detect binding started threads.